willow/plugins/pillow.py

```diff
@@ -246,4 +254,144 @@ def to_buffer_rgba(self):
         return RGBAImageBuffer(image.size, image.tobytes())
 
 
-willow_image_classes = [PillowImage]
+class PillowAnimatedImage(Image):
+    def __init__(self, frames):
+        self.frames = frames
+
+    @classmethod
+    def check(cls):
+        _PIL_Image()
+
+    @classmethod
+    def is_format_supported(cls, image_format):
+        return is_format_supported(image_format)
+
+    @Image.operation
+    def get_size(self):
+        return self.frames[0].get_size()
+
+    @Image.operation
+    def get_frame_count(self):
+        return len(self.frames)
+
+    @Image.operation
+    def has_alpha(self):
+        return self.frames[0].has_alpha()
+
+    @Image.operation
+    def has_animation(self):
+        return self.get_frame_count() > 1
+
+    @Image.operation
+    def resize(self, size):
+        return PillowAnimatedImage([frame.resize(size) for frame in self.frames])
+
+    @Image.operation
+    def crop(self, rect):
+        return PillowAnimatedImage([frame.crop(rect) for frame in self.frames])
+
+    @Image.operation
+    def rotate(self, angle):
+        return PillowAnimatedImage([frame.rotate(angle) for frame in self.frames])
+
+    @Image.operation
+    def set_background_color_rgb(self, color):
+        return PillowAnimatedImage([frame.set_background_color_rgb(color) for frame in self.frames])
+
+    @Image.operation
+    def save_as_jpeg(self, f, quality=85, optimize=False, progressive=False):
+        if self.has_animation():
+            pass  # TODO: Raise warning
+
+        return self.frames[0].save_as_jpeg(f, quality=quality, optimize=optimize, progressive=progressive)
+
+    @Image.operation
+    def save_as_png(self, f, optimize=False):
+        if self.has_animation():
+            pass  # TODO: Raise warning
+
+        return self.frames[0].save_as_png(f, optimize=optimize)
+
+    @Image.operation
+    def save_as_gif(self, f):
+        image = self.frames[0].image
+        frames = self.frames
+
+        # All gif files use either the L or P mode but we sometimes convert them
+        # to RGB/RGBA to improve the quality of resizing. We must make sure that
+        # they are converted back before saving.
+        if image.mode not in ['L', 'P']:
+            frames = [
```

Comment:
for an RGBA image, e.g. https://upload.wikimedia.org/wikipedia/commons/thumb/4/47/%2BFinger-und-N%C3%A4gel-einer-Studentin-in-der-Mensa---Bild-001.gif/1280px-%2BFinger-und-N%C3%A4gel-einer-Studentin-in-der-Mensa---Bild-001.gif this fails with `'PillowImage' object has no attribute 'convert'` because the frames are `PillowImage` instances, rather than images.

Something like
```python
            converted_frames = []
            for frame in frames:
                frame.image = frame.image.convert('P', palette=_PIL_Image().ADAPTIVE)
                converted_frames.append(frame)

            frames = converted_frames
```

should do. This'll need a test

Reply:
Worth noting that I got "unrecognized raw mode" with Pillow 4.0 when trying to edit the uploaded image. Upgraded to  5.4.1 and it worked. I believe it is something related to https://github.com/python-pillow/Pillow/pull/2374

Reply:
The above test image consumes ~35M memory to generate the rendition shown on its edit page

Reply:
This should be solved now that `frames` is a list of `PIL.Image` objects.

I'll just add the test now.